<a href="https://colab.research.google.com/github/namithashaji/California-Housing-Analysis/blob/main/California_Housing_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# California Housing Analysis

Dataset:  
Use the California Housing dataset available in the sklearn library. This dataset contains
information about various features of houses in California and their respective median prices.

## 1. Loading and Preprocessing

 Load the California Housing dataset using the fetch_california_housing function
from sklearn.  
● Convert the dataset into a pandas DataFrame for easier handling. Handle
missing values (if any) and perform necessary feature scaling (e.g.,
standardization).  
● Explain the preprocessing steps you performed and justify why they are
necessary for this dataset.

In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, MinMaxScaler, LabelEncoder, OneHotEncoder
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn import metrics
from sklearn.datasets import fetch_california_housing

In [ ]:
housing= fetch_california_housing() #Loading dataset
x= pd.DataFrame(housing.data, columns=housing.feature_names) #converting input dataset to DataFrame for easy handling.6
y=pd.Series(housing.target) #converting target dataset to Series
x.head(5)

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25


In [ ]:
x.info() #checking info about dataset. This give basic information including no. of entries, features and its dtype, count of non-null values.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   MedInc      20640 non-null  float64
 1   HouseAge    20640 non-null  float64
 2   AveRooms    20640 non-null  float64
 3   AveBedrms   20640 non-null  float64
 4   Population  20640 non-null  float64
 5   AveOccup    20640 non-null  float64
 6   Latitude    20640 non-null  float64
 7   Longitude   20640 non-null  float64
dtypes: float64(8)
memory usage: 1.3 MB


In [ ]:
x.isna().sum() #checking for null values. This improves quality of dataset.

,0
MedInc,0
HouseAge,0
AveRooms,0
AveBedrms,0
Population,0
AveOccup,0
Latitude,0
Longitude,0


In [ ]:
x.duplicated().sum() #checking for duplicate values. This also increases quality of dataset.

np.int64(0)

In [ ]:
standardscaler= StandardScaler() #feature scaling using StandardScaler method. This standardizes dataset's independent variables within a certain range.
x_standardscaled= standardscaler.fit_transform(x)

In [ ]:
x_standardscaled

array([[ 2.34476576,  0.98214266,  0.62855945, ..., -0.04959654,
         1.05254828, -1.32783522],
       [ 2.33223796, -0.60701891,  0.32704136, ..., -0.09251223,
         1.04318455, -1.32284391],
       [ 1.7826994 ,  1.85618152,  1.15562047, ..., -0.02584253,
         1.03850269, -1.33282653],
       ...,
       [-1.14259331, -0.92485123, -0.09031802, ..., -0.0717345 ,
         1.77823747, -0.8237132 ],
       [-1.05458292, -0.84539315, -0.04021111, ..., -0.09122515,
         1.77823747, -0.87362627],
       [-0.78012947, -1.00430931, -0.07044252, ..., -0.04368215,
         1.75014627, -0.83369581]])

#2. Regression Algorithm Implementation
● Implement the following regression algorithms:

○ Linear Regression  
○ Decision Tree Regressor  
● For each algorithm: Provide a brief explanation of how it works. Explain why it
might be suitable for this dataset.  

In [ ]:
x_train, x_test, y_train, y_test= train_test_split(x_standardscaled, y, test_size=0.2) #Splitting dataset as train data and test data in 80:20 ratio.

In [ ]:
x_standardscaled.shape #no. of rows and columns in dataset

(20640, 8)

In [ ]:
x_train.shape #no. of rows and columns in training set

(16512, 8)

In [ ]:
x_test.shape #no. of rows and columns in test set

(4128, 8)

### Linear Regression

Linear Regression is a statistical model that assumes a linear relationship between the input features (independent variables) and the target variable (dependent variable). It tries to find the best-fitting straight line that minimizes the sum of squared differences between the predicted values and the actual values.

**Suitability for this dataset:** Linear Regression can be suitable for the California Housing dataset if there's an assumption that the median house value (target) changes proportionally with the input features like median income, house age, etc. It's a good starting point for regression tasks and can identify which features have a statistically significant linear relationship with the target.

In [ ]:
linear_model= LinearRegression() #Implementation of Linear Regression.
linear_model.fit(x_train, y_train)

LinearRegression()

In [ ]:
linear_y_pred= linear_model.predict(x_test)
linear_y_pred

array([1.94023086, 1.75250887, 2.41448521, ..., 2.05651644, 2.25479448,
       0.64825019])

### Decision Tree Regressor

A Decision Tree Regressor builds a tree-like model of decisions, where each internal node represents a 'test' on an attribute, each branch represents the outcome of the test, and each leaf node represents the prediction value. The tree is built by recursively partitioning the data into subsets based on the features that best reduce impurity in the target variable.

**Suitability for this dataset:** Decision Tree Regressors are well-suited for the California Housing dataset because they can capture non-linear relationships and interactions between features without requiring explicit feature engineering for such relationships. They are robust to outliers and can handle both numerical and categorical data implicitly. This allows them to model complex decision rules that might dictate house prices, such as different pricing structures based on specific ranges of median income, number of rooms, or location (latitude/longitude).

In [ ]:
decission_tree= DecisionTreeRegressor() #Implementation of Decision Tree Regressor.
decission_tree.fit(x_train, y_train)

DecisionTreeRegressor()

In [ ]:
decission_y_pred= decission_tree.predict(x_test)
decission_y_pred

array([1.913, 1.729, 2.679, ..., 1.835, 1.375, 0.58 ])

#3. Model Evaluation and Comparison
● Evaluate the performance of each algorithm using the following metrics:  
○ Mean Squared Error (MSE)  
○ Mean Absolute Error (MAE)  
○ R-squared Score (R²)  
● Compare the results of all models and identify: The best-performing algorithm
with justification and the worst-performing algorithm with reasoning.

In [ ]:
linear_mae= metrics.mean_absolute_error(y_test, linear_y_pred)
linear_msa= metrics.mean_squared_error(y_test, linear_y_pred)
linear_r2= metrics.r2_score(y_test, linear_y_pred)
decission_tree_mae= metrics.mean_absolute_error(y_test, decission_y_pred)
decission_tree_msa= metrics.mean_squared_error(y_test, decission_y_pred)
decission_tree_r2= metrics.r2_score(y_test, decission_y_pred)
print("Performance Evaluation for Linear Regression \n--------------------------------------------")
print("Mean Absolute Error: ", linear_mae)
print("Mean Squared Error: ", linear_msa)
print("R2 Score: ", linear_r2)
print("\n")
print("Performance Evaluation for Decision Tree Regression \n--------------------------------------------")
print("Mean Absolute Error: ", decission_tree_mae)
print("Mean Squared Error: ", decission_tree_msa)
print("R2 Score: ", decission_tree_r2)

Performance Evaluation for Linear Regression 
--------------------------------------------
Mean Absolute Error:  0.5322499443463993
Mean Squared Error:  0.5518780203521931
R2 Score:  0.5789509202842302


Performance Evaluation for Decision Tree Regression 
--------------------------------------------
Mean Absolute Error:  0.4582373861434109
Mean Squared Error:  0.5160468624513809
R2 Score:  0.6062878960341606


###Based on the evaluation metrics calculated:

####Decision Tree Regressor is the best-performing algorithm.

* It has a lower Mean Absolute Error (MAE) of 0.4582 compared to Linear Regression's 0.5322. A lower MAE indicates better accuracy.
* It also has a lower Mean Squared Error (MSE) of 0.5160 compared to Linear Regression's 0.5518. A lower MSE indicates better performance.
* Furthermore, its R-squared Score (R²) of 0.6062 is higher than Linear Regression's 0.5789. A higher R² indicates a better fit of the model to the data.

The Decision Tree Regressor is likely performing better because it can capture non-linear relationships and interactions between features that a linear model cannot.


####Linear Regression is the worst-performing algorithm between the two.

* It consistently shows higher MAE and MSE, and a lower R² score compared to the Decision Tree Regressor.
* This suggests that the linear model might not be complex enough to capture the underlying patterns in the California Housing dataset, which often exhibit non-linear relationships between features and housing prices.